go/adbc/driver/databricks/cmd_record_reader.go

```diff
@@ -0,0 +1,187 @@
+		var arrowType arrow.DataType
+		switch col["type"] {
+		case "string":
+			arrowType = arrow.BinaryTypes.String
+		case "int":
+			arrowType = arrow.PrimitiveTypes.Int32
+		case "long":
+			arrowType = arrow.PrimitiveTypes.Int64
+		case "float":
+			arrowType = arrow.PrimitiveTypes.Float32
+		case "double":
+			arrowType = arrow.PrimitiveTypes.Float64
+		case "boolean":
+			arrowType = arrow.FixedWidthTypes.Boolean
+		case "timestamp":
+			arrowType = arrow.FixedWidthTypes.Timestamp_us
+		case "date":
+			arrowType = arrow.FixedWidthTypes.Date32
+		default:
+			arrowType = arrow.BinaryTypes.String
+		}
+		fields[i] = arrow.Field{
+			Name:     col["name"].(string),
+			Type:     arrowType,
+			Nullable: true,
+		}
+	}
+	return arrow.NewSchema(fields, nil)
+}
+
+func NewCommandRecordReader(
+	cmdExecution compute.CommandExecutionInterface, commandId string, result *compute.Results) (*cmdReader, error) {
+	// Convert the Databricks schema to an Arrow schema
+	schema := DeriveSchema(result.Schema)
+	r := &cmdReader{
+		refCount: 1,
+
+		cmdExecution: cmdExecution,
+
+		CommandId:     commandId,
+		CommandResult: result,
+
+		schema: schema,
+		rec:    nil,
+		err:    nil,
+		rec_read: false,
+
+		cancelFn: func() {},
+
+		startTime:     time.Now(),
+		BytesReceived: 1,
+		WaitTime:      0,
+	}
+
+	// For command execution, we need to convert the result to an Arrow record
+	if result.ResultType == compute.ResultTypeTable {
+
+		switch result.Data.(type) {
+		case []interface{}:
+			rows := result.Data.([]interface{})
+			r.rec, r.err = BuildFromRows(schema, rows)
+		default:
+			r.err = adbc.Error{
+				Code: adbc.StatusInternal,
+				Msg:  fmt.Sprintf("Unexpected command result type: %T", result.Data),
+			}
+		}
+
+	} else if result.ResultType == compute.ResultTypeText {
+		// For text result, return an empty record with a single string column
+		fields := []arrow.Field{{Name: "text", Type: arrow.BinaryTypes.String, Nullable: true}}
+		schema := arrow.NewSchema(fields, nil)
+		rows := result.Data.([]interface{})
+		r.rec, r.err = BuildFromRows(schema, rows)
+	} else {
+		r.err = adbc.Error{
+			Code: adbc.StatusInternal,
+			Msg:  fmt.Sprintf("Unexpected command result type: %s", result.ResultType),
+		}
+	}
+
+	return r, nil
+}
+
+// \post: if returns true, r.Record() != nil && r.err == nil
+// \post: if returns false, r.Record() == nil and r.err *MUST* be checked
+func (r *cmdReader) Next() bool {
+	if r.rec_read {
+		return false
+	}
+	r.rec_read = true
+	return true
+}
+func (r *cmdReader) Record() arrow.Record {
+	return r.rec
+}
+
+func (r *cmdReader) Err() error {
+	return r.err
+}
+
+func (r *cmdReader) Retain() {
+	atomic.AddInt64(&r.refCount, 1)
+}
+
+func (r *cmdReader) Release() {
+	if atomic.AddInt64(&r.refCount, -1) == 0 {
+		if r.rec != nil {
+			r.rec.Release()
+		}
+		// TODO: cancel HTTP connection
+		// TODO: close channel
+		r.cancelFn()
+	}
+}
+
+func (r *cmdReader) TotalRowCount() int64 {
+	return r.rec.NumRows()
```

Comment:
When you move the reading logic to here, you can get this number from the `*results` pointer that the reader is supposed to read Arrow record batches from. This makes it possible to call `TotalRowCount` before the record batch is produced.